internal/dispatch/remote/cluster.go

```diff
@@ -295,50 +339,47 @@ func dispatchSyncRequest[Q requestMessage, S responseMessage](
 			resp, err := handler(withTimeout, cr.clusterClient)
 			log.Trace().Str("request-key", reqKey).Msg("primary dispatch completed")
 			primaryResultChan <- respTuple[S]{resp, err}
-			hedgeWaitHistogram.WithLabelValues(reqKey).Observe(computedWait.Seconds())
 			primaryDispatch.WithLabelValues("false", reqKey).Inc()
 		}
 	}()
 
-	log.Trace().Object("request", req).Msg("running dispatch expression")
-	result, err := RunDispatchExpr(expr, req)
-	if err != nil {
-		log.Warn().Err(err).Msg("error when trying to evaluate the dispatch expression")
-	} else {
-		for _, secondaryDispatchName := range result {
-			secondary, ok := cr.secondaryDispatch[secondaryDispatchName]
-			if !ok {
-				log.Warn().Str("secondary-dispatcher-name", secondaryDispatchName).Msg("received unknown secondary dispatcher")
-				continue
-			}
+	for _, secondaryDispatchName := range secondaryDispatcherNames {
+		secondary, ok := cr.secondaryDispatch[secondaryDispatchName]
+		if !ok {
+			log.Warn().Str("secondary-dispatcher-name", secondaryDispatchName).Msg("received unknown secondary dispatcher")
+			continue
+		}
 
-			go func() {
-				select {
-				case <-withTimeout.Done():
-					log.Trace().Str("secondary", secondary.Name).Msg("secondary dispatch timed out or was canceled")
-					return
+		go func() {
+			select {
+			case <-withTimeout.Done():
+				log.Trace().Str("secondary", secondary.Name).Msg("secondary dispatch timed out or was canceled")
+				return
 
-				default:
-					log.Trace().Str("secondary", secondary.Name).Msg("running secondary dispatch")
-					startTime := time.Now()
-					resp, err := handler(withTimeout, secondary.Client)
-					endTime := time.Now()
-					handlerDuration := endTime.Sub(startTime)
-					if err != nil {
-						// For secondary dispatches, ignore any errors, as only the primary will be handled in
-						// that scenario.
-						log.Trace().Stringer("duration", handlerDuration).Str("secondary", secondary.Name).Err(err).Msg("got ignored secondary dispatch error")
-						cr.supportedResourceSubjectTracker.updateForError(err)
-						return
-					}
+			default:
+				log.Trace().Str("secondary", secondary.Name).Msg("running secondary dispatch")
+				startTime := time.Now()
+				resp, err := handler(withTimeout, secondary.Client)
+				endTime := time.Now()
+
+				// Cancel the primary sleeper as soon as any result comes back.
+				primarySleeper.cancelSleep()
```

Comment:
Why do we cancel the sleeper unconditionally? If the secondary has returned a legit result, the primary could be executed before **the result is sent over the channel**, which is what cancels the context. As a consequence, it will trigger an unnecessary primary request.

Please note that `primarySleeper.cancelSleep` does not cancel `withTimeout` context: the `sleep` method is creating a new context, and thus that cancellation is not propagated.

You could solve this by passing the top-level context and the cancellation function to the `primarySleeper` helper instead of creating a new context.